### PR TITLE
Fix directory aliasing for iTunes

### PR DESCRIPTION
**Pull request by hacker-cb**

### CODE REVIEW

**Comment by Be-ing** [comment]
By the way, there is no need to open a new pull request when you add commits. When you push commits to a branch that you have an open pull request for, the new commits will be added to the already open pull request.

**Comment by Hacker-CB** [comment]
Ok, sorry. I will do such in future.

**Comment by daschuer** [comment]
Cool thank you. I hope I find time to test this soon. 

In order to merge this, we need your permission. 
Please sign: 
https://docs.google.com/a/mixxx.org/spreadsheet/viewform?formkey=dEpYN2NkVEFnWWQzbkFfM0ZYYUZ5X2c6MQ
and comment here when done. 

Does it also solve your duplication issue because of your symlink? 

If I get the layout right you have something like this: 

Mixx Library on:
home/user/Musik

Symlink 
home/user/Musik/WindowsK -> /media/user/K

Right? (It is the same on my machine)

If you now play a file from the iTunes feature, it might be possible to add it as duplicate on 
/media/user/K/file.mp3 instead of the desired home/user/Musik/WindowsK/file.mp3

I have the same layout, unfortunately I have not uses iTunes since long. 
But I am going to test this. 

Mybe this is related https://bugs.launchpad.net/mixxx/+bug/1650633

If you like, you can fix a code style issue: Please surround = with spaces.

File: src/library/itunes/itunesfeature.cpp [hunk 1]
```diff
@@ -713,6 +735,10 @@ void ITunesFeature::parsePlaylist(QXmlStreamReader &xml, QSqlQuery &query_insert
                 //qDebug() << "exit playlist";
                 break;
             }
+            if (xml.name() == "dict" && !isPlaylistItemsStarted){
+                // Some playlists can be empty, so we need to exit.
+                break;
```
Comment (daschuer) [inline src/library/itunes/itunesfeature.cpp]: Why is this required? 
Without testing, I would say it has to continue to parse the remaining crates.

File: src/library/itunes/itunesfeature.cpp [hunk 2]
```diff
@@ -370,12 +374,27 @@ TreeItem* ITunesFeature::importLibrary() {
...
         }
     }
 
     itunes_file.close();
+    
+    if (isMusicFolderLocatedAfterTracks) {
+      qDebug() << "Updating iTunes real path from " << m_dbItunesRoot << " to " << m_mixxxItunesRoot;
+      // In some iTunes files "Music Folder" XML node is located at the end of file. So, we need to 
+      QSqlQuery query(m_database);
+      query.prepare("UPDATE itunes_library SET location = replace( location, :itunes_path, :mixxx_path )");
```
Comment (daschuer) [inline src/library/itunes/itunesfeature.cpp]: It should be safer to use "ltrim"

Comment (Hacker-CB) [inline src/library/itunes/itunesfeature.cpp]: 1. ltrim() [search and remove specified characters](https://www.techonthenet.com/sqlite/functions/ltrim.php), not substring. So, only replace() can be used if we want to make it in one query. In other case we need to iterate over each row.
2. QSting.replace() also called in original replacement:
````
if (!m_dbItunesRoot.isEmpty()) {
    location.replace(m_dbItunesRoot, m_mixxxItunesRoot);
}

**Comment by Hacker-CB** [comment]
- Duplication issue also solved by this fix. There is no duplicated files in main Library, after playing some item from iTunes.
- I have fixed code style and joined agreement.
- I think my fix will not affect https://bugs.launchpad.net/mixxx/+bug/1650633

**Comment by daschuer** [comment]
I have finally managed to test it wit iTunes 7 and 12. Both are working nice. 
LGTM. Thank you very much.

**Comment by daschuer** [comment]
We list all contributors inside Mixxx at the about box.
Normally we use our real names there. How should I list you?

**Comment by Hacker-CB** [comment]
You can use my real name -  Pavel Sokolov.
Thank you.


Daniel Schürmann <notifications@github.com> 14 января 2017 г. 3:31:40 написал:
